arraycontext/impl/pytato/__init__.py

```diff
@@ -0,0 +1,212 @@
+                    " arrays")
+        if isinstance(array, cla.Array):
+            return array.with_queue(None)
+        if not isinstance(array, pt.Array):
+            raise TypeError("PytatoPyOpenCLArrayContext.freeze invoked with "
+                            f"non-pytato array of type '{type(array)}'")
+
+        prg = pt.generate_loopy(array, cl_device=self.queue.device)
+        evt, (cl_array,) = prg(self.queue)
+        evt.wait()
+
+        return cl_array.with_queue(None)
+
+    def thaw(self, array):
+        import pytato as pt
+        import pyopencl.array as cla
+
+        if not isinstance(array, cla.Array):
+            raise TypeError("PytatoPyOpenCLArrayContext.thaw expects CL arrays, got "
+                    f"{type(array)}")
+
+        return pt.make_data_wrapper(array.with_queue(self.queue))
+
+    # }}}
+
+    def compile(self, f: Callable[..., Any]) -> Callable[..., Any]:
+        from arraycontext.impl.pytato.compile import PytatoCompiledOperator
+        return PytatoCompiledOperator(self, f)
+
+    def transform_loopy_program(self, prg):
+        from loopy.translation_unit import for_each_kernel
+
+        nwg = 48
+        nwi = (16, 2)
+
+        @for_each_kernel
+        def gridify(knl):
+            # {{{ Pattern matching inames
+
+            for insn in knl.instructions:
+                if isinstance(insn, lp.CallInstruction):
+                    # must be a callable kernel, don't touch.
+                    pass
+                elif isinstance(insn, lp.Assignment):
+                    bigger_loop = None
+                    smaller_loop = None
+                    for iname in insn.within_inames:
+                        if iname.startswith("iel"):
+                            assert bigger_loop is None
+                            bigger_loop = iname
+                        if iname.startswith("idof"):
+                            assert smaller_loop is None
+                            smaller_loop = iname
+
+                    if bigger_loop or smaller_loop:
+                        assert bigger_loop is not None and smaller_loop is not None
+                    else:
+                        sorted_inames = sorted(tuple(insn.within_inames),
+                                key=knl.get_constant_iname_length)
+                        smaller_loop = sorted_inames[0]
+                        bigger_loop = sorted_inames[1]
+
+                    knl = lp.chunk_iname(knl, bigger_loop, nwg,
+                            outer_tag="g.0")
+                    knl = lp.split_iname(knl, f"{bigger_loop}_inner",
+                            nwi[0], inner_tag="l.1")
+                    knl = lp.split_iname(knl, smaller_loop,
+                            nwi[1], inner_tag="l.0")
+                elif isinstance(insn, lp.BarrierInstruction):
+                    pass
+                else:
+                    raise NotImplementedError
+
+            # }}}
```

Comment:
#50 deprecated this on the pyopencl front, this logic really belongs in meshmode. (And it should use the transform metadata introduced in https://github.com/inducer/meshmode/pull/236 instead of matching names.)

Reply:
Done in 430c8d8